### PR TITLE
Switch to SHA256 and generate a nonce for AES operations

### PR DESCRIPTION
**Pull request by aaronjwood**
#118

### CODE REVIEW

**Comment by leonklingele** [comment]
Wow, this is just bad. ;) Calling `SetSuperSecureCookie` twice with the same key will reuse the same nonce + key combination, making the system utterly broken.

File: context.go [hunk 1]
```diff
@@ -417,11 +419,22 @@ func (ctx *Context) GetSecureCookie(key string) (string, bool) {
...
-	m := md5.Sum([]byte(secret))
-	secret = hex.EncodeToString(m[:])
-	text, err := com.AESEncrypt([]byte(secret), []byte(value))
+
+	// Ideally we'd want to use PBKDF2 here.
+	// Stick with SHA256 for now so we don't break the API.
+	// This method would need to take in a salt and pass it to PBKDF2 if we want to switch to PBKDF2.
+	hash := sha256.Sum256([]byte(secret))
+	secret = hex.EncodeToString(hash[:])
+	text, err := com.AESEncrypt([]byte(secret), ctx.nonce, []byte(value)) // The nonce should really be unique per cookie.
```
Comment (leonklingele) [inline context.go]: Wait. You hash the secret, encode it to a HEX string, then use the byte repesentation of that HEX string?
Why not plug in hash directly?

**Comment by aaronjwood** [comment]
@leonklingele I agree, we really need to switch to PBKDF2 and change the API to fix the issues you're mentioning. If the API can't be changed and we can't move to PBKDF2 then I don't know what else to do.

File: context.go (hunk 1 above)
Comment (aaronjwood) [inline context.go]: It should be as you say, unless there are dependencies on other libraries/packages that are expecting it in this way. Then they'd all need to be changed. @Unknwon would know more about that than me.

**Comment by leonklingele** [comment]
What you just said doesn't make sense at all. You're already breaking the API by switching from MD5 to SHA256. Then why not do it properly (i.e. without HEX encoding / decoding)?

**Comment by aaronjwood** [comment]
@leonklingele I don't think this breaks the API. It's internal to `Context`. If someone upgrades the framework then there would be all new cookies getting set, no? Also, I'm not adding all of the hex encoding/decoding, it was already there. If @Unknwon doesn't want to change the API then maybe this should just be closed.

**Comment by leonklingele** [comment]
It breaks the API because everyone who still has one of the old MD5-stuff can't use it anymore.

**Comment by aaronjwood** [comment]
@Unknwon I think it's best to close this out if the API can't be changed. I don't think what I've done here really helps anything. At the very least you can keep the changes to work with the new AES `nonce` parameter.

**Comment by leonklingele** [comment]
Why? IMO it's fine to break the API if the old one was broken security-wise. But then please do it properly.

**Comment by aaronjwood** [comment]
@Unknwon how about changing the API and have something in the next release notes about a bunch of flawed crypto getting fixed which causes incompatibility?

**Comment by Unknwon** [comment]
OK, I agree to break SetSuperSecureCookie's API as long as it allows accept arbitrary length of slat, and use it to generate desired nonce.

The other big use case of this function is https://github.com/grafana/grafana/tree/master/vendor/gopkg.in/macaron.v1, but it is in vendor so shouldn't break their build. But we should warn them (maybe file an issue) about this.

**Comment by aaronjwood** [comment]
Okay, let me close this out and open up a new, proper one.

**Pull request closed without merging**